src/Libraries/Nop.Services/Helpers/UserAgentHelper.cs

```diff
@@ -49,7 +54,33 @@ public virtual bool IsSearchEngine()
                     return false;
 
                 var userAgent = _httpContext.Request.UserAgent;
-                return _browscapXmlHelper.IsCrawler(userAgent);
+                if (String.IsNullOrEmpty(userAgent))
+                    return false;
+
+                // check in already found cache
+                bool result;
+                if (_foundSearchEnginesCache.TryGetValue(userAgent, out result))
+                    return true;
+
+                // check cookie
+                var cookie = _httpContext.Request.Cookies[IS_SEARCH_ENGINE_REQUEST_COOKIE_KEY];
+                if (cookie != null)
+                    return cookie.Value == "1";
+
+                // check browscap
+                result = _browscapXmlHelper.IsCrawler(userAgent);
+
+                // if IS search engine, save to cache
+                if (result)
+                    _foundSearchEnginesCache.TryAdd(userAgent, true);

```

Comment:
This approach problem is that  it becomes possible to organize a simple attack, the result of which will be the exhaustion of all available RAM and closing nopCommerce with the OutOfMemory exception